## Gaussian Elimination

### 1. Pivoting vs. No Pivoting

a. Using gaussian elimination, rounding to three decimal places each step..

In [68]:
def linearsolver(A,b):
    n = len(A)
    M = A

    i = 0
    for x in M:
        x.append(b[i])
        i += 1

    for k in range(n):
        for i in range(k,n):
            if abs(M[i][k]) > abs(M[k][k]):
                M[k], M[i] = M[i],M[k]
            else:
                pass

    for j in range(k+1,n):
        q = float(M[j][k]) / M[k][k]
        for m in range(k, n+1):
            M[j][m] -=  q * M[k][m]

    x = [0 for i in range(n)]

    x[n-1] =float(M[n-1][n])/M[n-1][n-1]
    for i in range (n-1,-1,-1):
        z = 0
        for j in range(i+1,n):
            z = z  + float(M[i][j])*x[j]
    x[i] = float(M[i][n] - z)/M[i][i]
    print(x)




In [89]:
import numpy as np

j=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
l=[11.2,-4.3,-.605,4.415]
print('initial matrix-  \n ' + str(np.asmatrix([j,k,l])))
j=[round((1/.143)*i,3) for i in j]
k=[round((1/-1.31)*i,3) for i in k]
l=[round((1/11.2)*i,3) for i in l]
print("now we have 1's in left most column -  \n " + str(np.asmatrix([j,k,l])))
k=np.subtract(k,j)
l=np.subtract(l,j)
print("R2-R1 , R3-R1   \n " + str(np.asmatrix([j,k,l])))
k=[round((1/-3.192)*i,3) for i in k]
l=[round((1/-2.881)*i,3) for i in l]

print("continuing with 2nd column   \n " + str(np.asmatrix([j,k,l])))
l=np.subtract(l,k)
l=[round((1/.019)*i,3) for i in l]
print( "... \n" + str(np.asmatrix([j,k,l])) + " \n now we have all 1's on diagonal so we will start back substitution \n ")

l_alpha=[round((-4.879)*i,3) for i in l]
k=np.add(k,l_alpha)
l_beta=[round((-14.056)*i,3) for i in l]
j=np.add(j,l_beta)
print("R2 + -.4879*R3   \n  R1 + -14.056*R3 " + str(np.asmatrix([j,k,l])))
k_alpha=[round((-2.497)*i,3) for i in k]
j=np.add(j,k_alpha)
print("R1 + -2.497*R2   \n " + str(np.asmatrix([j,k,l])))

initial matrix-  
 [[  0.143   0.357   2.01   -5.173]
 [ -1.31    0.911   1.99   -5.458]
 [ 11.2    -4.3    -0.605   4.415]]
now we have 1's in left most column -  
 [[  1.      2.497  14.056 -36.175]
 [  1.     -0.695  -1.519   4.166]
 [  1.     -0.384  -0.054   0.394]]
R2-R1 , R3-R1   
 [[  1.      2.497  14.056 -36.175]
 [  0.     -3.192 -15.575  40.341]
 [  0.     -2.881 -14.11   36.569]]
continuing with 2nd column   
 [[  1.      2.497  14.056 -36.175]
 [ -0.      1.      4.879 -12.638]
 [ -0.      1.      4.898 -12.693]]
... 
[[  1.      2.497  14.056 -36.175]
 [ -0.      1.      4.879 -12.638]
 [  0.      0.      1.     -2.895]] 
 now we have all 1's on diagonal so we will start back substitution 
 
R2 + -.4879*R3   
  R1 + -14.056*R3 [[ 1.     2.497  0.     4.517]
 [-0.     1.     0.     1.487]
 [ 0.     0.     1.    -2.895]]
R1 + -2.497*R2   
 [[ 1.     0.     0.     0.804]
 [-0.     1.     0.     1.487]
 [ 0.     0.     1.    -2.895]]


This gives solution [0.804, 1.487, -2.895] which is different from the actual solution..
<br\>
b. using pivoting

In [102]:
import numpy as np

j=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
l=[11.2,-4.3,-.605,4.415]

print('initial matrix-  \n ' + str(np.asmatrix([j,k,l])) + 
      '\n we need to swap the pivot row (R3) with R1  \n ')

l=[0.143,0.357,2.01,-5.173]
k=[-1.31,.911,1.99,-5.458]
j=[11.2,-4.3,-.605,4.415]

print(" ... \n " + str(np.asmatrix([j,k,l])))

j=[round((1/11.2)*i,3) for i in j]
k=[round((1/-1.31)*i,3) for i in k]
l=[round((1/.143)*i,3) for i in l]

print(" \n getting 1's in left most column -  \n " + str(np.asmatrix([j,k,l])))
k=np.subtract(k,j)
l=np.subtract(l,j)
print("subtracting first row form second and third-   \n " + str(np.asmatrix([j,k,l])) +
      "\n before continuing with 2nd column, need to swap R2 and R3 \n ")

a=k
b=l
l=a
k=b

print("...   \n " + str(np.asmatrix([j,k,l])))

k=[round((1/2.881)*i,3) for i in k]
l=[round((1/-0.311)*i,3) for i in l]

print("continuing with 2nd column  \n " + str(np.asmatrix([j,k,l])))
l=np.subtract(l,k)
l=[round((1/-0.187)*i,3) for i in l]
print( "... \n" + str(np.asmatrix([j,k,l])) + " \n now we have all 1's on diagonal so we will start back substitution \n ")

l_alpha=[round((-4.898)*i,3) for i in l]
k=np.add(k,l_alpha)
l_beta=[round((.054)*i,3) for i in l]
j=np.add(j,l_beta)
print("R2 + -4.898*R3   \n  R1 + .054*R3 \n" + str(np.asmatrix([j,k,l])))
k_alpha=[round((0.384)*i,3) for i in k]
j=np.add(j,k_alpha)
print("R1 + 0.384*R2   \n " + str(np.asmatrix([j,k,l])))

initial matrix-  
 [[  0.143   0.357   2.01   -5.173]
 [ -1.31    0.911   1.99   -5.458]
 [ 11.2    -4.3    -0.605   4.415]]
 we need to swap the pivot row (R3) with R1  
 
 ... 
 [[ 11.2    -4.3    -0.605   4.415]
 [ -1.31    0.911   1.99   -5.458]
 [  0.143   0.357   2.01   -5.173]]
 
 getting 1's in left most column -  
 [[  1.     -0.384  -0.054   0.394]
 [  1.     -0.695  -1.519   4.166]
 [  1.      2.497  14.056 -36.175]]
subtracting first row form second and third-   
 [[  1.     -0.384  -0.054   0.394]
 [  0.     -0.311  -1.465   3.772]
 [  0.      2.881  14.11  -36.569]]
 before continuing with 2nd column, need to swap R2 and R3 
 
...   
 [[  1.     -0.384  -0.054   0.394]
 [  0.      2.881  14.11  -36.569]
 [  0.     -0.311  -1.465   3.772]]
continuing with 2nd column  
 [[  1.     -0.384  -0.054   0.394]
 [  0.      1.      4.898 -12.693]
 [ -0.      1.      4.711 -12.129]]
... 
[[  1.     -0.384  -0.054   0.394]
 [  0.      1.      4.898 -12.693]
 [  0.     -0.      1.    

Here we have a much better solution [1.029,2.079,-3.016], notice the multipliers in the back substitution were much smaller (the goal of pivoting) which causes less change in the augmented vector which in turn causes less error due to rounding.

## 2. Ill-Conditioned Systems

Consider the linear system with

$$
Ax = b
$$

with

$$
A =
\begin{bmatrix}
1& 1 \\
1& 1.001
\end{bmatrix}
$$

and $b$ arbitrary. Note that $A$ is very close to being non-invertible, a sure sign of ill-conditioned systems (small perturbations in $b$ can lead to large perturbations in the solution).

### (a) Calculate the condition number of $A$ using the max norm for matrices.

Max norm is just a the max row sum of $A$.

The condition number is less than or equal to $||A||*||A^{-1}||$

Which is:

$2.001 * 2001 = 4002.001$

### (b)

Intuition:

The solution to y represents the case where we perturb b by a small amount. Because this system is relatively ill-conditioned, then the difference between x and y should be much larger than the magnitude of the perturbance.

Demostration:

Let $b = \begin{bmatrix}1 \\ 0 \end{bmatrix}$ and $\epsilon = \begin{bmatrix}0.01 \\ -0.01 \end{bmatrix}$


Then $x = A^{-1} * b = \begin{bmatrix}1001 \\ -1000 \end{bmatrix}$

and $y = A^{-1} * (b + \epsilon) = \begin{bmatrix}1021.01 \\ -1020 \end{bmatrix}$

now $||y-x||_{\infty} = 20.01$ and $||\epsilon||_{\infty} = .01$. Clearly this is a large difference, and supports the behavior expected by my intuition.


# Iterative Methods for Linear Systems
## 3. Approximating Solutions to Integral Equations

## 4. From Divergence to Convergence for Iterative Methods

Consider the system

$$
Ax = b
$$

with

$$
A = \begin{bmatrix} 
5& 1& 0 \\
10& -1& 1 \\
-15& -4& 4
\end{bmatrix}
, b = \begin{bmatrix}2 \\ -3 \\ 5 \end{bmatrix}
$$

### (a) Show that Gauss-Jacobi will diverge.

Gauss-Jacobi will diverge if the spectral radius of M (the method for producing matrix M from matrix A is defined in the text) is greater than 1.

The following code will demonstrate that M's spectral radius is greater than 1, and also demonstrate the divergence of Guasse-Jacobi.

In [37]:
import numpy as np

def gauss_jacobi_step(A, b, x):
    x_next = np.copy(x)
    for r in range(0, A.shape[0]):
        row = A[r]
        a_ii = row[r]
        b_i = b[r]
        x_i = x[r]
        
        x_next[r] = (b_i - (np.dot(row, x) - (a_ii * x_i))) / a_ii
    return x_next

def div_rows_by_diag(A):
    A_new = np.copy(A)
    for r in range(0, A.shape[0]):
        row = A[r]
        row = -1 * row / row[r]
        row[r] = 0.
        A_new[r] = row
    return A_new
    

In [40]:
A = np.array([[5., 1., 0.],[10., -1., 1.],[-15., -4., 4.]])
b = np.array([2., -3., 5.])
M = div_rows_by_diag(A)

print("Eigenvalues of M:")
eigvals = np.linalg.eigvals(M)
real_eigvals = eigvals[eigvals.imag == 0.]
print(real_eigvals)

print("Observe Gauss-Jacobi diverging")
x = np.array([0., 0., 0.])
x_last = x
for i in range(0, 10):
    print(np.linalg.norm(x - x_last))
    x_last = x
    x = gauss_jacobi_step(A, b, x)

Eigenvalues of M:
[-0.56736423+0.j]
Observe Gauss-Jacobi diverging
0.0
3.2745228660065884
6.940641180755565
3.5146123541579946
9.268570884985452
6.987991932594084
9.2080637215432
13.313427564516395
7.165033501535425
19.44137446312953


### (b) Use elementary row operations to convert the system to an equivalent system that is diagonally dominant

$$
Begin: \begin{bmatrix} 
5& 1& 0 \\
10& -1& 1 \\
-15& -4& 4
\end{bmatrix}
$$

$$
R2 = R2 - 2*R1 \begin{bmatrix} 
5& 1& 0 \\
0& -3& 1 \\
-15& -4& 4
\end{bmatrix}
$$

$$
R3 = R3 + 3*R1 \begin{bmatrix} 
5& 1& 0 \\
0& -3& 1 \\
0& -1& 4
\end{bmatrix}
$$

$$
\bar{A} = \begin{bmatrix} 
5& 1& 0 \\
0& -3& 1 \\
0& -1& 4
\end{bmatrix}
$$

We will show diagonal dominance by transforming $\bar{A}$ into $\bar{M}$, performing a row sum, and asserting that the final vector has no element > 1.

In [41]:
A_bar = np.array([[5., 1., 0.],[0., -3., 1.],[0., -1., 4.]])
b = np.array([2., -3., 5.])
M_bar = div_rows_by_diag(A_bar)
np.sum(M_bar, axis=1)


array([-0.2       ,  0.33333333,  0.25      ])

## 6. Eigenvalues and Eigenfunctions of an Integral Operator

* Each row of the matrix should correspond with each approximation of $v''(t_i)$
* Each column of the matrix should correspond with each $v(t_i)$ component in the approximation of $v''(t_i)$
* Vector $v$ is unknown (this will be an eigenvector)
* $\lambda$ is unknown (this will be the eigenvalues)
* By construction, $v_0 = v(0) = v(1) = v_n(1) = 0$, so the last column of the matrix will represent both the 0th and the nth component.
* There should be $n$ columns and $n$ rows
* On the nth row, the $i+1$ component should be equal to the 1st component b/c $v$ is periodic

$$
A = \begin{bmatrix}
\frac{-2}{h^2}& \frac{1}{h^2}& 0& 0& ... & 0 \\
\frac{1}{h^2}& \frac{-2}{h^2}& \frac{1}{h^2}& 0& ... & 0 \\
0& \frac{1}{h^2}& \frac{-2}{h^2}& \frac{1}{h^2}& ... & 0 \\
...& ...& ...& ...& ...& ...
\end{bmatrix}
$$

In [53]:
import numpy as np

def qr_method(A):
    n_cols = A.shape[1]
    Q = np.zeros((n_cols, n_cols))
    
    Q[:,0] = A[:,0]
    
    for k in range(0, n_cols - 1):
        Q[:,k+1] = A[:,k+1]
        
        for i in range(0, k):
            dot_quotient = np.dot(A[:,k+1], Q[:,i]) / np.dot(Q[:,i], Q[:,i])
            Q[:,k+1] = Q[:,k+1] - dot_quotient * Q[:,i]
            
    for i in range(0, n_cols):
        Q[:,i] = Q[:,i] / np.linalg.norm(Q[:,i])
        
        if Q[i, i] < 0:
            Q[:,i] = -1 * Q[:,i]
            
    return Q

def inverse_iter(A, b, l):
    lI = np.zeros(A.shape)
    np.fill_diagonal(lI, l)
    
    b = np.linalg.inv(A - lI) @ b
    
    return b / np.linalg.norm(b)

def build_array(n):
    A = np.zeros((n, n))
    h = 1. / n
    h_2 = np.full(n, 1. / h**2)
    
    np.fill_diagonal(A, -2. / h**2)
    np.fill_diagonal(A[1:], h_2)
    np.fill_diagonal(A[:,1:], h_2)
    
    A[n - 1, 0] = h_2[0]
    return A
    

In [55]:
A = build_array(20)
E = np.copy(A)
for i in range(0, 10):
    Q = qr_method(E)
    Qt = Q.transpose()
    E = Qt*E*Q

eigenvalues = np.diagonal(E)
last_4_eigvals = np.sort(eigenvalues)[-4:]

b = np.ones(20)
for i in range(0, 100):
    b = inverse_iter(A, b, last_4_eigvals[0])
    
b

b = np.ones(20)
for i in range(0, 100):
    b = inverse_iter(A, b, last_4_eigvals[1])
    
b

b = np.ones(20)
for i in range(0, 100):
    b = inverse_iter(A, b, last_4_eigvals[2])
    
b

b = np.ones(20)
for i in range(0, 100):
    b = inverse_iter(A, b, last_4_eigvals[1])
    
b


array([-3.00750478e-01, -1.85874017e-01,  1.85874017e-01,  3.00750478e-01,
       -1.15735543e-16, -3.00750478e-01, -1.85874017e-01,  1.85874017e-01,
        3.00750478e-01, -1.52890147e-16, -3.00750478e-01, -1.85874017e-01,
        1.85874017e-01,  3.00750478e-01, -1.85674635e-16, -3.00750478e-01,
       -1.85874017e-01,  1.85874017e-01,  3.00750478e-01, -2.45621820e-16])